File: domain/src/main/java/com/ahn/domain/favorite/usecase/BuildFavoriteRatesUseCase.kt
package com.ahn.domain.favorite.usecase

import com.ahn.domain.currency.model.CurrencyInfo
import com.ahn.domain.favorite.model.FavoriteRateInfo
import javax.inject.Inject

class BuildFavoriteRatesUseCase @Inject constructor() {
    operator fun invoke(
        baseCurrency: CurrencyInfo?,
        baseAmount: String,
        favoriteCurrencyCodes: List<String>,
        availableCurrencies: List<CurrencyInfo>,
        ratesByCode: Map<String, Double>,
    ): List<FavoriteRateInfo> {
        if (
            baseCurrency == null ||
            favoriteCurrencyCodes.isEmpty() ||
            availableCurrencies.isEmpty()
        ) {
            return emptyList()
        }

        val baseAmountValue = baseAmount.toDoubleOrNull() ?: 0.0
        val currenciesByCode = availableCurrencies.associateBy { it.code }

        return favoriteCurrencyCodes.distinct().mapNotNull { code ->
            if (code == baseCurrency.code) return@mapNotNull null

            val currency = currenciesByCode[code] ?: return@mapNotNull null
            val rate = ratesByCode[code] ?: return@mapNotNull null

            FavoriteRateInfo(
                currency = currency,
                baseCurrencyCode = baseCurrency.code,
                rate = rate,
                convertedAmount = baseAmountValue * rate,
            )
        }
    }
}

File: ios/Packages/Domain/Sources/Domain/Favorite/UseCase/BuildFavoriteRatesUseCase.swift
public struct BuildFavoriteRatesUseCase: Sendable {
    public init() {}

    public func callAsFunction(
        baseCurrency: CurrencyInfo?,
        baseAmount: String,
        favoriteCurrencyCodes: [String],
        availableCurrencies: [CurrencyInfo],
        ratesByCode: [String: Double],
    ) -> [FavoriteRateInfo] {
        guard
            let baseCurrency,
            !favoriteCurrencyCodes.isEmpty,
            !availableCurrencies.isEmpty
        else {
            return []
        }

        let baseAmountValue = Double(baseAmount) ?? 0.0
        let currenciesByCode = Dictionary(
            availableCurrencies.map { ($0.code, $0) },
            uniquingKeysWith: { _, last in last },
        )

        var seenCodes = Set<String>()

        return favoriteCurrencyCodes.compactMap { code in
            guard seenCodes.insert(code).inserted else { return nil }
            guard code != baseCurrency.code else { return nil }
            guard let currency = currenciesByCode[code], let rate = ratesByCode[code] else { return nil }

            return FavoriteRateInfo(
                currency: currency,
                baseCurrencyCode: baseCurrency.code,
                rate: rate,
                convertedAmount: baseAmountValue * rate,
            )
        }
    }
}
